pkg/agent/openflow/pipeline_windows.go

```diff
@@ -190,11 +200,78 @@ func (c *client) snatImplementationFlows(nodeIP net.IP, category cookie.Category
 	return flows
 }
 
+// serviceSNATFlows installs flows to allow a Windows host to access a ClusterIP Service.
+func (c *client) serviceSNATFlows(nodeIP net.IP, localGatewayMAC net.HardwareAddr, category cookie.Category) []binding.Flow {
+	var flows []binding.Flow
+	if c.enableProxy {
+		flows = append(flows, []binding.Flow{
+			// If the host uses another IP to access the Service, hostnetwork ClusterIP Service reply packet
+			// needs unSNAT with snatCTMark mark.
+			c.pipeline[uplinkTable].BuildFlow(priorityNormal).
+				MatchProtocol(binding.ProtocolIP).
+				MatchRegRange(int(marksReg), markTrafficFromUplink, trafficFromRegRange).
+				Action().CT(false, unSNATTable, ctZoneSNAT).NAT().CTDone().
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				Done(),
+			c.pipeline[unSNATTable].BuildFlow(priorityNormal).
+				MatchProtocol(binding.ProtocolIP).
+				MatchCTMark(snatCTMark, nil).
+				MatchCTStateRpl(true).
+				MatchCTStateTrk(true).
+				Action().SetDstMAC(localGatewayMAC).
+				Action().ResubmitToTable(conntrackTable).
+				Done(),
+			c.pipeline[unSNATTable].BuildFlow(priorityLow).
+				MatchRegRange(int(marksReg), markTrafficFromUplink, trafficFromRegRange).
+				Action().Output(config.BridgeOFPort).
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				Done(),
+			// Flows below are for hostnetwork ClusterIP Service which needs SNAT and the SNAT mark is set.
+			c.pipeline[l3ForwardingTable].BuildFlow(priorityLow).MatchProtocol(binding.ProtocolIP).
+				MatchInPort(config.HostGatewayOFPort).MatchCTMark(ServiceCTMark, nil).
+				MatchRegRange(int(marksReg), markTrafficFromGateway, trafficFromRegRange).
+				Action().LoadRegRange(int(marksReg), snatDefaultMark, snatMarkRange).
+				Action().ResubmitToTable(l2ForwardingCalcTable).
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				Done(),
+			// SNAT: antrea-gw0 -> NodeIP.
+			c.pipeline[conntrackCommitTable].BuildFlow(priorityHigh).MatchProtocol(binding.ProtocolIP).
+				MatchCTMark(ServiceCTMark, nil).
+				MatchCTStateNew(true).MatchCTStateTrk(true).
+				MatchRegRange(int(marksReg), snatDefaultMark, snatMarkRange).
+				MatchInPort(config.HostGatewayOFPort).
+				Action().CT(true, L2ForwardingOutTable, ctZoneSNAT).SNAT(
+				&binding.IPRange{StartIP: nodeIP, EndIP: nodeIP},
+				nil).
+				LoadToMark(snatCTMark).CTDone().
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				Done(),
+			c.pipeline[conntrackCommitTable].BuildFlow(priorityHigh).MatchProtocol(binding.ProtocolIP).
+				MatchCTMark(ServiceCTMark, nil).
+				MatchCTStateNew(false).MatchCTStateTrk(true).
+				MatchInPort(config.HostGatewayOFPort).
+				MatchRegRange(int(marksReg), snatDefaultMark, snatMarkRange).
+				Action().CT(false, L2ForwardingOutTable, ctZoneSNAT).NAT().CTDone().
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				Done(),
+			// The request packet is received from antrea-gw0 and will be sent back to antrea-gw0 for routing, so
```

Comment:
This flow is for hostnetwork case and I think tunnel is for no-hostnetwork.
I am not sure why you mentioned "local pod" since this PR supports traffic from Windows host rather than a pod. If you refer to the case that endpoint is on the same Windows node as the request, still it is a hostnetwork case and Windows doesn't support hostnetwork pod.

Reply:
For the hostnetwork case here, I think the traffic should be:
Request:
process --> antrea-gw0 --> DNAT --> SNAT --> antrea-gw0 --> host --> br-int --> Uplink --> RemoteNode
Reply:
RemoveNode --> Uplink --> unSNAT --> unDNAT --> antrea-gw0 --> process

Reply:
Thanks for the clarification! I think my confusion is due to I misunderstood the earlier flow in l3FwdTable and thought it will catch all host->Service packets, but I now see it is of low-priority, so only packets to external will be caught.

But since the host -> hostNetwork/external endpoint case is pretty complicated, would you mind add the path to comments in line 229?

Reply:
Updated.